library.go

```diff
@@ -0,0 +1,147 @@
+	e, err := transport.NewEndpoint(endpoint)
+	if err != nil {
+		return "", err
+	}
+
+	if !strings.HasSuffix(e.Path, ".git") {
+		e.Path += ".git"
+	}
+
+	return RepositoryID(path.Join(e.Host, e.Path)), nil
+}
+
+func (id RepositoryID) String() string {
+	return string(id)
+}
+
+// Mode is the different modes to open a Repository.
+type Mode int
+
+const (
+	// RWMode allows to perform read and write operations over a repository.
+	RWMode Mode = iota
+	// ReadOnlyMode allows only read-only operations over a repository.
+	ReadOnlyMode
+)
+
+// Repository interface represents a git.Repository, with information about
+// how it was open and where is located. Also provides Transactional
+// capabilities through the method Commit.
+type Repository interface {
+	// ID returns the RepositoryID.
+	ID() RepositoryID
+	// LocationID returns the LocationID from the Location where it was retrieved.
+	LocationID() LocationID
+	// Mode returns the Mode how it was opened.
+	Mode() Mode
+	// Commit persists all the write operations done since was open, if the
+	// repository doesn't provide Transactional capabilities should return
+	// ErrNonTransactional.
+	Commit() error
+	// Close closes the repository, if the repository was opened in transactional
+	// Mode, will delete any write operation pending to be written.
+	Close() error
+	// R returns the git.Repository.
+	R() *git.Repository
+}
+
+// LibraryID represents a Library identifier.
+type LibraryID string
+
+// Library interface represents a group of different libraries and locations,
+// it allows access to any repository stored on any library or location. Also
+// allows the iteration of the libraries and locations to perform full scan
+// operations. Library is the default entrypoint for accessing the repositories,
+// should be used when the Location is not important.
+type Library interface {
+	// ID returns the LibraryID for this Library.
+	ID() LibraryID
+	// Init initializes a new Repository in a Location, the choosen Location
+	// is dependant on the implementation, if this this not supported should
+	// return ErrNotImplemented. If a repository with the given RepositoryID
+	// already exists ErrRepositoryExists is returned.
+	Init(RepositoryID) (Repository, error)
+	// Get open a repository with the given RepositoryID, it itereates all the
+	// library locations until this repository is found. If a repository with
+	// the given RepositoryID can't be found the ErrRepositoryNotExists is
+	// returned.
+	Get(RepositoryID, Mode) (Repository, error)
+	// GetOrInit open or initilizes a Repository at a Location, if this this not
+	// supported should return ErrNotImplemented. If the repository is opened
+	// this will be done in RWMode.
+	GetOrInit(RepositoryID) (Repository, error)
+	// Has returns true, the LibraryID and the LocationID if the given
+	// RepositoryID matches any repository at any location belonging to this
+	// Library.
+	Has(RepositoryID) (bool, LibraryID, LocationID, error)
```

Comment:
Worth to return the `Library` and `Location` instead of just the ids?

Reply:
I believe so. I believe in most cases we will use the objects instead of their IDs.

Reply:
You believe this? I think that the object isn't that useful because most likely you will have already the location, or you only need the IDs for partitioning

Reply:
For me it makes sense to have IDs as the interface for the library user to manage the abstractions, so if you really want a certain object you should request it explicitly through the ID instead of having a method `Has` returning objects which maybe you don't need.

On the other hand, it feels weird having a method `Has(Repo)` returning all these things, maybe `Has` should only do what its name suggests and return `(bool, error)` and add other method to retrieve the coordinates `(LibraryId, LocationID)`

Reply:
Since doing `Has` its expensive, doesn't have a sense of having two different methods. Also, if you call Has and Get, you are traversing all the inner Libraries and Locations searching for it twice. 

I think that makes sense the name `Has` since it answers that has it and where is it. 

About having the ID and not the object, means that you need to call ID function may be for nothing, and after that, you need to search inside of the inner libraries for the object itself, when you already have it.

Reply:
> Since doing Has its expensive, doesn't have a sense of having two different methods. Also, if you call Has and Get, you are traversing all the inner Libraries and Locations searching for it twice.

If the `Has` operation is expensive or not it's a problem part of the specific implementation and should be solved there instead of make it reflected on the API design. Anyway, maybe this is not very important so :+1: